File: src/com/suushiemaniac/cubing/alglib/move/modifier/SquareOneHalfModifier.java
package com.suushiemaniac.cubing.alglib.move.modifier;

public class SquareOneHalfModifier implements Modifier {
    public static SquareOneHalfModifier fromNumber(int number) {
        SquareOneDirectionModifier directionModifier = number < 0 ? SquareOneDirectionModifier.NEG : SquareOneDirectionModifier.POS;
        SquareOneNumModifier numModifier = SquareOneNumModifier.fromNotation(Math.abs(number) + "");
        return new SquareOneHalfModifier(directionModifier, numModifier);
    }

    private SquareOneDirectionModifier directionModifier;
    private SquareOneNumModifier numModifier;

    public SquareOneHalfModifier(SquareOneDirectionModifier directionModifier, SquareOneNumModifier numModifier) {
        this.directionModifier = directionModifier;
        this.numModifier = numModifier;
    }

    @Override
    public SquareOneHalfModifier inverse() {
        return new SquareOneHalfModifier(this.directionModifier.inverse(), this.numModifier);
    }

    @Override
    public SquareOneHalfModifier merge(Modifier other) {
        if (!(other instanceof SquareOneHalfModifier)) return null;
        int newRot = this.toNumber() + ((SquareOneHalfModifier) other).toNumber();
        if (newRot < -6) newRot = 6 + (newRot + 6);
        else if (newRot > 6) newRot = -6 + (newRot - 6);
        return SquareOneHalfModifier.fromNumber(newRot);
    }

    @Override
    public String toFormatString() {
        return this.directionModifier.toFormatString() + this.numModifier.toFormatString();
    }

    public int toNumber() {
        return Integer.parseInt(this.toFormatString());
    }

    public SquareOneDirectionModifier getDirectionModifier() {
        return this.directionModifier;
    }

    public SquareOneNumModifier getNumModifier() {
        return this.numModifier;
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/move/modifier/SquareOneHalfModifier.kt
package com.suushiemaniac.cubing.alglib.move.modifier

import kotlin.math.abs

data class SquareOneHalfModifier(val directionModifier: SquareOneDirectionModifier, val numModifier: SquareOneNumModifier) : Modifier {
    override val notation = "${this.directionModifier.notation}${this.numModifier.notation}"

    override fun inverse() = SquareOneHalfModifier(this.directionModifier.inverse(), this.numModifier)

    override fun merge(other: Modifier): SquareOneHalfModifier {
        if (other !is SquareOneHalfModifier) return other as SquareOneHalfModifier

        val newRot = (this.toNumber() + other.toNumber()).let {
            when {
                it < -6 -> 6 + (it + 6)
                it > 6 -> -6 + (it - 6)
                else -> it
            }
        }

        return SquareOneHalfModifier.fromNumber(newRot)
    }

    fun toNumber() = this.notation.toInt()
    override fun toString() = this.notation

    companion object {
        fun fromNumber(number: Int): SquareOneHalfModifier {
            val directionModifier = if (number < 0) SquareOneDirectionModifier.NEG else SquareOneDirectionModifier.POS
            val numModifier = SquareOneNumModifier.values().find { it.notation == abs(number).toString() }!!

            return SquareOneHalfModifier(directionModifier, numModifier)
        }
    }
}
